src/Servers/Kestrel/Core/src/Internal/WebTransport/WebTransportSession.cs

```diff
@@ -0,0 +1,201 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+using System.Net.Http;
+using System.Threading.Channels;
+using Microsoft.AspNetCore.Connections;
+using Microsoft.AspNetCore.Connections.Features;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3;
+using Microsoft.AspNetCore.Server.Kestrel.Core.WebTransport;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.WebTransport;
+
+/// <summary>
+/// Controls the WebTransport session and keeps track of all the streams.
+/// </summary>
+internal class WebTransportSession : IWebTransportSession
+{
+    private readonly CancellationTokenRegistration _connectionClosedRegistration;
+
+    // stores all created streams (pending or accepted)
+    private readonly ConcurrentDictionary<long, WebTransportStream> _openStreams = new();
+    // stores all pending streams that have not been accepted yet
+    private readonly Channel<WebTransportStream> _pendingStreams;
+
+    private readonly Http3Connection _connection;
+    private readonly Http3Stream _connectStream = default!;
+    private bool _isClosing;
+
+    private static readonly ReadOnlyMemory<byte> OutputStreamHeader = new(new byte[] {
+            0x40 /*quic variable-length integer length*/,
+            (byte)Http3StreamType.WebTransportUnidirectional,
+            0x00 /*body*/});
+
+    internal static string WebTransportProtocolValue => "webtransport";
+    internal static string SecPrefix => "sec-webtransport-http3-";
+    internal static string VersionHeaderPrefix => $"{SecPrefix}draft";
+    internal static string CurrentSuppportedVersion => $"{VersionHeaderPrefix}02";
+
+    long IWebTransportSession.SessionId => _connectStream.StreamId;
+
+    internal WebTransportSession(Http3Connection connection, Http3Stream connectStream)
+    {
+        _connection = connection;
+        _connectStream = connectStream;
+        _isClosing = false;
+        // unbounded as limits to number of streams is enforced elsewhere
+        _pendingStreams = Channel.CreateUnbounded<WebTransportStream>();
+
+        // listener to abort if this connection is closed
+        _connectionClosedRegistration = connection._multiplexedContext.ConnectionClosed.Register(state =>
```

Comment:
```suggestion
        _connectionClosedRegistration = connection._multiplexedContext.ConnectionClosed.Register(static state =>
```

Reply:
There could be multiple sessions at a time (ie. from different connections and due to having multiple sessions per connection), wouldn't this static cause issues with these states?

Reply:
You're passing in the state to the `Register` call. Static here is just a way of us avoiding accidentally allocating a lambda closure when we know we are passing in the needed state.